### PR TITLE
Feature Flag + Mobile Sync

### PR DESCRIPTION
**Pull request by danfinlay**
Replaces #5935.

Hides that feature (which doesn't make sense until we have mobile) behind a hidden feature flag.

Adds the ability to set feature flags from the background console using this method:

```javascript
/*
* @method
* Sets a feature flag to the `enabled` value.
* @param { String } featureName - The unique name for the feature.
* @param { Boolean } enabled - Whether the feature should be enabled.
*/
function setPreference (featureName, enabled) { /* implementation */ }
```

You can try it out by entering `setPreference('mobileSync', true)` in the background console, which will add a new button to the bottom of the settings view.

### CODE REVIEW

File: ui/app/components/pages/mobile-sync/index.js [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+
+  generateCipherKeyAndChannelName () {
+    this.cipherKey = `${this.props.selectedAddress.substr(-4)}-${PubNub.generateUUID()}`
+    this.channelName = `mm-${PubNub.generateUUID()}`
+  }
+
+  initWebsockets () {
+    this.pubnub = new PubNub({
+      subscribeKey: 'sub-c-30b2ba04-c37e-11e8-bd78-d63445bede87',
+      publishKey: 'pub-c-d40e77d5-5cd3-4ca2-82eb-792a1f4573db',
```
Comment (brunobar79) [inline ui/app/components/pages/mobile-sync/index.js]: We gotta rotate keys before merging this in, move the keys into CircleCI env variables and consume it here via process.env.PUBNUB_SUB_KEY and process.env.PUBNUB_PUB_KEY

File: app/scripts/metamask-controller.js [hunk 2]
```diff
@@ -583,6 +587,66 @@ module.exports = class MetamaskController extends EventEmitter {
...
+    const hdKeyring = this.keyringController.getKeyringsByType('HD Key Tree')[0]
+    const hdAccounts = await hdKeyring.getAccounts()
+    const accounts = {
+      hd: hdAccounts.filter((item, pos) => (hdAccounts.indexOf(item) === pos)).map(address => ethUtil.toChecksumAddress(address)),
+      simpleKeyPair: [],
+      ledger: [],
+      trezor: [],
+    }
+    /*
+    const simpleKeyPairKeyring = await this.keyringController.getKeyringsByType('Simple Key Pair')[0]
```
Comment (brunobar79) [inline app/scripts/metamask-controller.js]: remove

**Review by brunobar79 — DISMISSED** [review]
Adding this comments so I can address before we decide to merge this in

**Comment by danfinlay** [comment]
There's an important item to note with this feature: Since it hangs a method on the `global` namespace, preferences managed are highly exposed to all dependencies, so we should make sure that none of the preferences have any potential for abuse by a malicious dependency.

**Comment by danfinlay** [comment]
Ok I've pushed up some comments to help encourage correct future use of the feature flags hash. I'm ok with this being merged if someone approves it now.

File: app/scripts/controllers/preferences.js [hunk 3]
```diff
@@ -33,7 +35,15 @@ class PreferencesController {
...
-      featureFlags: {},
+
+      // WARNING: Do not use feature flags for security-sensitive things.
+      // Feature flag toggling is available in the global namespace
+      // for convenient testing of pre-release features, and should never
+      // perform sensitive operations.
+      featureFlags: {
+        betaUI: true,
+        skipAnnounceBetaUI: true,
+      },
```
Comment (whymarrh) [inline app/scripts/controllers/preferences.js]: We should drop this addition before merging as this seems to be a regression

**Review by whymarrh — CHANGES_REQUESTED** [review]
Two small things:

1. The noted regression (?) in the PreferencesController
2. The changes to the `package-lock.json` file should probably be dropped and re-added, I'm seeing this PR downgrade the version of json-rpc-engine we're pinned to from 3.8.0 to 3.7.3 which doesn't seem correct.

**Review by whymarrh — APPROVED** [review]
Looks fine to me

**Comment by guix77** [comment]
@danfinlay on MM 6.5.3 (FF + Chrome) under Settings > Advanced, the mobile sync is not hidden any more. Does not help with confusion about the "release" of the mobile app medias talked about and malware fake apps :)

**Pull request merged**